### PR TITLE
Redirect user profile pages

### LINKED ISSUE #6915: Redirect user profile pages (only public pages, not edit pages): username => user ID
See also: #6526

---

After #6526, all user profile pages are accessed using the user ID in the URL and not the `username` anymore. All the links generated by AMO use the user ID too, but we should support the old URLs (i.e. the URLs containing `username`s).

⚠️ we will not redirect "edit" profile URLs because those URLs are not public/published.

**For QA:** make sure that loading a profile page with a `username` in the URL redirects to the same profile page but with the user ID in the URL.

### PR DESCRIPTION
**Pull request by willdurand**
Fixes #6915
~~:warning: depends on #6904~~

---

This PR allows to keep backward compatibility with the previous URL scheme. User profiles loaded _via_ the `username` in the URL will be redirected to the URL with the user ID.

I double checked with product: we do not need to setup a redirect for the "edit" pages (private pages).

I also inlined a few comments for context. There is #6769 for the `componentWillReceiveProps()` but the logic in this component is tough...

### CODE REVIEW

File: src/amo/pages/UserProfile/index.js [hunk 1]
```diff
@@ -173,7 +196,7 @@ export class UserProfileBase extends React.Component<InternalProps> {
 
     invariant(currentUser, 'currentUser is required');
 
-    if (currentUser.id === params.userId) {
+    if (`${currentUser.id}` === params.userId) {
```
Comment (willdurand) [inline src/amo/pages/UserProfile/index.js]: The test case was not realistic. `params.userId` is a string and `===` checks the type too 🤒

File: src/amo/pages/UserProfileEdit/index.js [hunk 2]
```diff
@@ -722,6 +722,7 @@ export class UserProfileEditBase extends React.Component<Props, State> {
               </p>
 
               <UserProfileEditNotifications
+                key={user && user.id}
```
Comment (willdurand) [inline src/amo/pages/UserProfileEdit/index.js]: This is an edge case, I noticed this component does not update the UI when switching from one user to another (which is super tricky to reproduce), even if all the props/data are correct.

File: src/amo/reducers/users.js [hunk 3]
```diff
@@ -362,7 +362,7 @@ export const loadUserNotifications = ({
 };
 
 export const getUserById = (users: UsersState, userId: UserId) => {
-  invariant(userId, 'userId is required');
+  invariant(typeof userId === 'number', 'userId is required');
```
Comment (willdurand) [inline src/amo/reducers/users.js]: This allows to support `0`, to get a 404 instead of a `500`.

File: tests/unit/amo/pages/TestUserProfile.js [hunk 4]
```diff
@@ -1029,6 +1031,32 @@ describe(__filename, () => {
     expect(root.find('meta[name="description"]')).toHaveLength(0);
   });
 
+  it('sends a server redirect when user profile is loaded with a "username" in the URL', () => {
+    const clientApp = CLIENT_APP_FIREFOX;
+    const lang = 'fr';
+    const userId = 1;
+    const { store } = dispatchSignInActions({ clientApp, lang, userId });
```
Comment (bobsilverberg) [inline tests/unit/amo/pages/TestUserProfile.js]: Why do we need to sign in with a different user? Wouldn't the test still work if the current user was the same as the user whose profile we're trying to view? As long as we use a username and not userId for the `userId` param it seems like the test would still be valid.

File: src/amo/pages/UserProfile/index.js [hunk 5]
```diff
@@ -101,7 +119,12 @@ export class UserProfileBase extends React.Component<InternalProps> {
       dispatch(
         fetchUserAccount({
           errorHandlerId: errorHandler.id,
-          userId: Number(params.userId),
+          // We should use `Number()` here but we need to fetch users by
+          // username in order to send a server redirect (to support previous
+          // URLs with usernames). That is why we have to ignore the Flow error
+          // here.
+          // $FLOW_IGNORE
+          userId: params.userId,
```
Comment (bobsilverberg) [inline src/amo/pages/UserProfile/index.js]: This adds the ability to fetch a user based on a string username, in addition to a numeric userid. Do you think it's worth adding a test with a string username to the suite for this? I notice that there isn't one in there.

File: tests/unit/amo/pages/TestUserProfile.js (hunk 4 above)
Comment (willdurand) [reply tests/unit/amo/pages/TestUserProfile.js]: Yep, that's right

File: src/amo/pages/UserProfile/index.js (hunk 1 above)
Comment (willdurand) [reply src/amo/pages/UserProfile/index.js]: I wasn't aware of this notation. I will use it!

File: src/amo/pages/UserProfile/index.js (hunk 5 above)
Comment (willdurand) [reply src/amo/pages/UserProfile/index.js]: There is a test case for that, no? https://github.com/mozilla/addons-frontend/blob/f3d02de14effb6a4d0c100fc9ae9afa3ab57c1c4/tests/unit/amo/pages/TestUserProfile.js#L1034

Comment (willdurand) [reply src/amo/pages/UserProfile/index.js]: Nevermind, I got it: you meant a test case to fetch a user, before the redirect.

**Pull request merged**